ext/windows/service/daemon.rb

```diff
@@ -175,7 +175,7 @@ def parse_runinterval(puppet_path)
 
   def parse_log_level(puppet_path,cmdline_debug)
     begin
-      loglevel = %x{ #{puppet_path} agent --configprint log_level}.chomp
+      loglevel = %x{ #{puppet_path} config --section agent print log_level }.chomp
```

Comment:
@joshcooper changing this line makes the Puppet Agent service not start... any idea on how to debug this? Nothing interesting in Event Viewer as well. Executing this on the CLI shows the same output as the deprecated command.

Reply:
This is working for me. What does  your `puppet.conf` look like?

In the Service Control Manager UI, you can pass startup arguments like `--logtofile --log_level debug`.

You might also try running `procmon` and filter on process name includes "ruby.exe" and see which Win32 calls are failing.

It's also possible to run the daemon as a foreground process, e.g. `ruby daemon.rb`, and pass arguments to it, but it's been awhile since I've done that.

Reply:
These are the puppet.conf contents
```
[main]
server=port-mainstream.delivery.puppetlabs.net
autoflush=true
manage_internal_file_permissions=false
log_level = debug
logdest = C:\puppet.txt
```
What I do is install the agent then run these two commands in the CLI:

```
C:\> puppet config set log_level debug --section main
C:\> puppet config set logdest C:\Users\Administrator --section main
```

After that, the service doesn't start. I passed `--logtofile --log_debug` to the daemon but nothing was logged. After reverting this line the service starts again. I also tried changing `--section agent` to `--section main` but that didn't help.

--------------------

Ok, with some logging inside daemon.rb to print the `loglevel` variable I think I got it:

```
loglevel: [0;36mDebug: Caching environment 'production' (ttl = 0 sec)[0m
[0;36mDebug: Evicting cache entry for environment 'production'[0m
[0;36mDebug: Deleted text domain :production: false[0m
[0;36mDebug: Caching environment 'production' (ttl = 0 sec)[0m
[0;36mDebug: Caching environment 'production' (ttl = 0 sec)[0m
[0;36mDebug: Runtime environment: puppet_version=6.15.0, ruby_version=2.5.8, run_mode=user, default_encoding=IBM437[0m
debug
```

It seems that the debug prints are also on stdout and it doesn't get parsed correctly. How should we circumvent this? The debug prints don't show when running `--configprint`. I assume this isn't a *bug* with `config print` since debug logging was explicitly set by us.

The simplest solution would be to `.split.last` on the output, and some checking on `loglevel`, to ensure what we receive is one of the supported log levels, otherwise fall back on a sane default (`:notice`).

Reply:
Ah, ok, it used to be that `log_level` couldn't be set in puppet.conf, so we never had issues with debug output appearing in the output. It also appears that `puppet agent --configprint` behaves differently than `puppet config print` when debug is enabled:

```
$ bx puppet agent --configprint runinterval
Warning: Using --configprint is deprecated. Use 'puppet config <subcommand>' instead.
   (location: /Users/josh/work/puppet/lib/puppet/settings.rb:501:in `print_config_options')
1800
```

```
$ bx puppet config print runinterval
Debug: Caching environment 'production' (ttl = 0 sec)
Debug: Evicting cache entry for environment 'production'
Debug: Deleted text domain :production: false
Debug: Caching environment 'production' (ttl = 0 sec)
Debug: Caching environment 'production' (ttl = 0 sec)
Debug: Runtime environment: puppet_version=6.15.0, ruby_version=2.5.7, run_mode=user, default_encoding=UTF-8
Resolving settings from section 'agent' in environment 'production'
1800
```

Note that `puppet agent --configprint` prints a deprecation warning to stderr, which is not captured by `%x{ }` in daemon.rb.

So one option is to always pass log_level on the command line (which overrides what's in puppet.conf)

```
$ bx puppet config print runinterval --log_level notice
1800
```

Reply:
> So one option is to always pass log_level on the command line (which overrides what's in puppet.conf)

I'll do this, I like this better than string parsing.